src/utils/gov.ts

```diff
@@ -0,0 +1,40 @@
+export const isWhitelisted = (url?: string) => {
+  if (!url) return false
+  if (!url.match(/https?:\/\/.*/)) {
+    url = `https://${url}`
+  }
+
+  /* common interchain proposal discussion forums */
+  const whitelistedDomains = [
+    "commonwealth.im",
+    "github.com",
+    "mintscan.io",
+    "terra.money",
+  ]
+
+  try {
+    const hostname = new URL(url).hostname
+    var whitelisted = false
+
+    whitelistedDomains.forEach((domain) => {
+      if (hostname === domain || hostname.endsWith(`.${domain}`)) {
+        whitelisted = true
+      }
+    })
+    return whitelisted
+  } catch (error) {
+    return false
+  }
+}
+
+/* exempted URLs are displayed as written (but not linked) */
+export const isExempted = (url?: string) => {
+  const exempted = ["terra.py"]
+  if (!url) return false
+
+  return exempted.includes(url.toLowerCase())
+}
+
+/* include 2 and 3 character domains + .money as anchor TLDs for url detection; based on https://data.iana.org/TLD/tlds-alpha-by-domain.txt */
+export const URL_REGEX =
+  /((?:https?:\/\/)?(?:[A-Z0-9-]+?\s*\.\s*)*?(?:\b[A-Z0-9-]{1,63}?)(?:(?:\s*\.(?:am|an|as|at|be|by|do|it|in|is|how|new|no|one|so|top|to))|(?:\s*\.\s*(?:ads|afl|app|axa|ac|ad|ae|af|ag|ai|al|ao|aq|ar|au|aw|ax|az|bar|bbc|bid|bio|biz|bmw|boo|bzh|ba|bb|bd|bf|bg|bh|bi|bj|bl|bm|bn|bo|bq|br|bs|bt|bv|bw|bz|cab|cal|cat|cbn|ceo|cfd|com|crs|ca|cc|cd|cf|cg|ch|ci|ck|cl|cm|cn|co|cr|cu|cv|cw|cx|cy|cz|dad|day|dev|dnp|de|dj|dk|dm|dz|eat|edu|esq|eus|ec|ee|eg|eh|er|es|et|eu|fan|fit|fly|foo|frl|fi|fj|fk|fm|fo|fr|gal|gdn|gle|gmo|gmx|goo|gop|gov|ga|gb|gd|ge|gf|gg|gh|gi|gl|gm|gn|gp|gq|gr|gs|gt|gu|gw|gy|hiv|hk|hm|hn|hr|ht|hu|ibm|ifm|ing|ink|int|iwc|id|ie|il|im|io|iq|ir|jcb|je|jm|jo|jp|kim|krd|ke|kg|kh|ki|km|kn|kp|kr|kw|ky|kz|lat|lds|la|lb|lc|li|lk|lr|ls|lt|lu|lv|ly|money|mil|mma|moe|mov|mtn|ma|mc|md|me|mf|mg|mh|mk|ml|mm|mn|mo|mp|mq|mr|ms|mt|mu|mv|mw|mx|my|mz|net|ngo|nhk|nra|nrw|ntt|nyc|na|nc|ne|nf|ng|ni|nl|np|nr|nu|nz|ong|onl|ooo|org|ovh|om|pro|pub|pa|pe|pf|pg|ph|pk|pl|pm|pn|pr|ps|pt|pw|py|qa|red|ren|rio|rip|re|ro|rs|ru|rw|sap|sca|scb|sew|sky|soy|sa|sb|sc|sd|se|sg|sh|si|sj|sk|sl|sm|sn|sr|ss|st|su|sv|sx|sy|sz|tax|tel|tui|tc|td|tf|tg|th|tj|tk|tl|tm|tn|tp|tr|tt|tv|tw|z|uno|uol|ua|ug|uk|um|us|uy|uz|vet|va|vc|ve|vg|vi|vn|vu|wed|win|wme|wtc|wtf|wf|ws|xin|xxx|xyz|ye|yt|zip|za|zm|zw))+?\b)(?:\/[^.!,?;"'<>()[\]{}\s\x7F-\xFF]*(?:[.!,?]+[^.!,?;"'<>()[\]{}\s\x7F-\xFF]+)*)?)/gi
```

Comment:
I've [ran some tests with this regex](https://regex101.com/r/plHDpz/1) and it seems to miss some cases:

* Valid TLD with more than three chars, e.g.`http://foo.bananarepublic`
* Valid TLD with non-ASCII signs, e.g. `http://foo.vermögensberater`
* TLD with spaces, e.g `http://foo.c om/bar`
* Valid domain-name that end with a non-ASCII sign, e.g. `http://foo🍔.com`

Especially the valid TLD and domain-names are critical, since most browsers will try to open the page directly when the text is pasted into the address-bar. The spaces are less important IMHO, because domains with spaces won't be recognized as a valid URL by the browsers I've tested, thus the page won't be loaded directly.

I therefore propose another regex (see below), which matches all valid domain-names and TLD. [As far as I have tested](https://regex101.com/r/VhEzt4/1) any string that would be recognized as an URL by a browser would be matched by the regex. However, this also comes at two expenses:

1. It is easier to trick the regex by inserting spaces. For example `http://foo . com/bar` won't be matched, while the original regex matches it. But I think the tradeoff is acceptable, because not allowing spaces around the dot prevents a lot of false positives and strings with spaces won't be recognized as URLs anyway.

2. The number of false-positively matched file-names, like *myscript.py*, has been increased, since also names like *myscript.java* are matched due to the increased number of TLDs. However, I problem also exists with the current regex and could be mitigated with a smart exemption-rule.

Finally, here is the new regular expression:

```
[^\s]+\.(?:zw|zuerich|zone|zm|zip|zero|zara|zappos|za|yun|yt|youtube|you|yokohama|yoga|yodobashi|ye|yandex|yamaxun|yahoo|yachts|xyz|xxx|xin|xihuan|xfinity|xerox|xbox|wtf|wtc|ws|wow|world|works|work|woodside|wolterskluwer|wme|winners|wine|windows|win|williamhill|wiki|wien|whoswho|wf|weir|weibo|wedding|wed|website|weber|webcam|weatherchannel|weather|watches|watch|wanggou|wang|walter|walmart|wales|vuelos|vu|voyage|voto|voting|vote|volvo|volkswagen|vodka|vn|vlaanderen|vivo|viva|vision|visa|virgin|vip|vin|villas|viking|vig|video|viajes|vi|vg|vet|versicherung|verisign|ventures|vegas|ve|vc|vanguard|vana|vacations|va|uz|uy|us|ups|uol|uno|university|unicom|uk|ug|ubs|ubank|ua|tz|tw|tvs|tv|tushu|tunes|tui|tube|tt|trv|trust|travelersinsurance|travelers|travelchannel|travel|training|trading|trade|tr|toys|toyota|town|tours|total|toshiba|toray|top|tools|tokyo|today|to|tn|tmall|tm|tl|tkmaxx|tk|tjx|tjmaxx|tj|tirol|tires|tips|tiffany|tienda|tickets|tiaa|theatre|theater|thd|th|tg|tf|teva|tennis|temasek|tel|technology|tech|team|tdk|td|tci|tc|taxi|tax|tattoo|tatar|tatamotors|target|taobao|talk|taipei|tab|sz|systems|sydney|sy|sx|swiss|swatch|sv|suzuki|surgery|surf|support|supply|supplies|sucks|su|style|study|studio|stream|store|storage|stockholm|stcgroup|stc|statefarm|statebank|star|staples|stada|st|ss|srl|sr|spot|sport|space|spa|soy|sony|song|solutions|solar|sohu|software|softbank|social|soccer|so|sncf|sn|smile|smart|sm|sling|sl|skype|sky|skin|ski|sk|sj|site|singles|sina|silk|si|showtime|show|shouji|shopping|shop|shoes|shiksha|shia|shell|shaw|sharp|shangrila|sh|sg|sfr|sexy|sex|sew|seven|services|sener|select|seek|security|secure|seat|search|se|sd|scot|science|schwarz|schule|school|scholarships|schmidt|schaeffler|scb|sca|sc|sbs|sbi|sb|saxo|save|sas|sarl|sap|sanofi|sandvikcoromant|sandvik|samsung|samsclub|salon|sale|sakura|safety|safe|saarland|sa|ryukyu|rwe|rw|run|ruhr|rugby|ru|rsvp|rs|room|rogers|rodeo|rocks|rocher|ro|rip|rio|ril|ricoh|richardli|rich|rexroth|reviews|review|restaurant|rest|republican|report|repair|rentals|rent|ren|reliance|reit|reisen|reise|rehab|redumbrella|redstone|red|recipes|realty|realtor|realestate|read|re|radio|racing|quest|quebec|qpon|qa|py|pwc|pw|pub|pt|ps|prudential|pru|protection|property|properties|promo|progressive|prof|productions|prod|pro|prime|press|praxi|pramerica|pr|post|porn|politie|poker|pohl|pnc|pn|pm|plus|plumbing|playstation|play|place|pl|pk|pizza|pioneer|pink|ping|pin|pid|pictures|pictet|pics|physio|photos|photography|photo|phone|philips|phd|pharmacy|ph|pg|pfizer|pf|pet|pe|pccw|pay|passagens|party|parts|partners|pars|paris|panasonic|page|pa|ovh|ott|otsuka|osaka|origins|organic|org|orange|oracle|open|ooo|online|onl|ong|one|omega|om|ollo|oldnavy|olayangroup|olayan|okinawa|office|observer|obi|nz|nyc|nu|ntt|nrw|nra|nr|np|nowtv|nowruz|now|norton|northwesternmutual|nokia|no|nl|nissay|nissan|ninja|nikon|nike|nico|ni|nhk|ngo|ng|nfl|nf|nexus|nextdirect|next|news|new|neustar|network|netflix|netbank|net|nec|ne|nc|nba|navy|natura|name|nagoya|nab|na|mz|my|mx|mw|mv|mutual|music|museum|mu|mtr|mtn|mt|msd|ms|mr|mq|mp|movie|mov|motorcycles|moto|moscow|mortgage|mormon|monster|money|monash|mom|moi|moe|moda|mobile|mobi|mo|mn|mma|mm|mls|mlb|ml|mk|mitsubishi|mit|mint|mini|mil|microsoft|miami|mh|mg|merckmsd|menu|men|memorial|meme|melbourne|meet|media|med|me|md|mckinsey|mc|mba|mattel|maserati|marshalls|marriott|markets|marketing|market|map|mango|management|man|makeup|maison|maif|madrid|macys|ma|ly|lv|luxury|luxe|lundbeck|lu|ltda|ltd|lt|ls|lr|lplfinancial|lpl|love|lotto|lotte|london|lol|locus|locker|loans|loan|llp|llc|lk|living|live|lipsy|link|linde|lincoln|limo|limited|lilly|like|lighting|lifestyle|lifeinsurance|life|lidl|li|lgbt|lexus|lego|legal|lefrak|leclerc|lease|lds|lc|lb|lawyer|law|latrobe|latino|lat|lasalle|lanxess|landrover|land|lancia|lancaster|lamer|lamborghini|lacaixa|la|kz|kyoto|ky|kw|kuokgroup|kred|krd|kr|kpn|kpmg|kp|kosher|komatsu|koeln|kn|km|kiwi|kitchen|kindle|kinder|kim|kids|kia|ki|kh|kg|kfh|kerryproperties|kerrylogistics|kerryhotels|ke|kddi|kaufen|juniper|juegos|jprs|jpmorgan|jp|joy|jot|joburg|jobs|jo|jnj|jmp|jm|jll|jio|jewelry|jetzt|jeep|je|jcb|java|jaguar|itv|itau|it|istanbul|ist|ismaili|is|irish|ir|iq|ipiranga|io|investments|intuit|international|int|insure|insurance|institute|ink|ing|info|infiniti|industries|inc|in|immobilien|immo|imdb|imamat|im|il|ikano|ifm|ieee|ie|id|icu|ice|icbc|ibm|hyundai|hyatt|hughes|hu|ht|hsbc|hr|how|house|hotmail|hotels|hoteles|hot|hosting|host|hospital|horse|honda|homesense|homes|homegoods|homedepot|holiday|holdings|hockey|hn|hm|hkt|hk|hiv|hitachi|hisamitsu|hiphop|hgtv|hermes|here|helsinki|help|healthcare|health|hdfcbank|hdfc|hbo|haus|hangout|hamburg|hair|gy|gw|guru|guitars|guide|guge|gucci|guardian|gu|gt|gs|group|grocery|gripe|green|gratis|graphics|grainger|gr|gq|gp|gov|got|gop|google|goog|goodyear|goo|golf|goldpoint|gold|godaddy|gn|gmx|gmo|gmbh|gmail|gm|globo|global|gle|glass|gl|giving|gives|gifts|gift|gi|gh|ggee|gg|gf|george|genting|gent|gea|ge|gdn|gd|gbiz|gb|gay|garden|gap|games|game|gallup|gallo|gallery|gal|ga|fyi|futbol|furniture|fund|fun|fujitsu|ftr|frontier|frontdoor|frogans|frl|fresenius|free|fr|fox|foundation|forum|forsale|forex|ford|football|foodnetwork|food|foo|fo|fm|fly|flowers|florist|flir|flights|flickr|fk|fj|fitness|fit|fishing|fish|firmdale|firestone|fire|financial|finance|final|film|fido|fidelity|fiat|fi|ferrero|ferrari|feedback|fedex|fast|fashion|farmers|farm|fans|fan|family|faith|fairwinds|fail|fage|extraspace|express|exposed|expert|exchange|events|eus|eurovision|eu|etisalat|et|estate|esq|es|erni|ericsson|er|equipment|epson|enterprises|engineering|engineer|energy|emerck|email|eg|ee|education|edu|edeka|eco|ec|eat|earth|dz|dvr|dvag|durban|dupont|dunlop|dubai|dtv|drive|download|dot|domains|dog|doctor|docs|do|dnp|dm|dk|dj|diy|dish|discover|discount|directory|direct|digital|diet|diamonds|dhl|dev|design|desi|dentist|dental|democrat|delta|deloitte|dell|delivery|degree|deals|dealer|deal|de|dds|dclk|day|datsun|dating|date|data|dance|dad|dabur|cz|cyou|cymru|cy|cx|cw|cv|cuisinella|cu|cruises|cruise|crs|crown|cricket|creditunion|creditcard|credit|cr|cpa|courses|coupons|coupon|country|corsica|coop|cool|cookingchannel|cooking|contractors|contact|consulting|construction|condos|comsec|computer|compare|company|community|commbank|comcast|com|cologne|college|coffee|codes|coach|co|cn|cm|clubmed|club|cloud|clothing|clinique|clinic|click|cleaning|claims|cl|ck|cityeats|city|citic|citi|citadel|cisco|circle|cipriani|ci|church|chrome|christmas|chintai|cheap|chat|chase|charity|channel|chanel|ch|cg|cfd|cfa|cf|cern|ceo|center|cd|cc|cbs|cbre|cbn|cba|catholic|catering|cat|casino|cash|case|casa|cars|careers|career|care|cards|caravan|car|capitalone|capital|capetown|canon|camp|camera|cam|calvinklein|call|cal|cafe|cab|ca|bzh|bz|by|bw|bv|buzz|buy|business|builders|build|bt|bs|brussels|brother|broker|broadway|bridgestone|bradesco|br|box|boutique|bot|boston|bostik|bosch|booking|book|boo|bond|bom|bofa|boehringer|boats|bo|bnpparibas|bn|bmw|bms|bm|blue|bloomberg|blog|blockbuster|blackfriday|black|bj|biz|bio|bingo|bing|bike|bid|bible|bi|bharti|bh|bg|bf|bet|bestbuy|best|berlin|bentley|beer|beauty|beats|be|bd|bcn|bcg|bbva|bbt|bbc|bb|bayern|bauhaus|basketball|baseball|bargains|barefoot|barclays|barclaycard|barcelona|bar|bank|band|bananarepublic|banamex|baidu|baby|ba|azure|az|axa|ax|aws|aw|avianca|autos|auto|author|auspost|audio|audible|audi|auction|au|attorney|athleta|at|associates|asia|asda|as|arte|art|arpa|army|archi|aramco|arab|ar|aquarelle|aq|apple|app|apartments|aol|ao|anz|anquan|android|analytics|amsterdam|amica|amfam|amex|americanfamily|americanexpress|amazon|am|alstom|alsace|ally|allstate|allfinanz|alipay|alibaba|alfaromeo|al|akdn|airtel|airforce|airbus|aig|ai|agency|agakhan|ag|africa|afl|af|aetna|aero|aeg|ae|adult|ads|ad|actor|aco|accountants|accountant|accenture|academy|ac|abudhabi|abogado|able|abc|abbvie|abbott|abb|abarth|aarp|aaa)(?:\/\S*)?(?=\s|$)
```

However, I would recommend to use the following optimized version, which needs significantly less steps for my test-case than the unoptimized version:

* optimized new regex: 1880 steps
* unoptimized new regex 41781 steps
* original regex: 13640 steps

```
[^\s]+\.(?:a(?:a(?:rp|a)|b(?:arth|b(?:ott|vie)|le|ogado|udhabi|[bc])|c(?:ademy|c(?:enture|ountants?)|tor|o)|d(?:ult|s)|e(?:ro|tna|g)|f(?:rica|l)|g(?:akhan|ency)|i(?:r(?:bus|force|tel)|g)|kdn|l(?:faromeo|i(?:baba|pay)|l(?:finanz|state|y)|s(?:ace|tom))|m(?:azon|e(?:rican(?:express|family)|x)|fam|ica|sterdam)|n(?:alytics|droid|quan|z)|ol|p(?:artments|ple|p)|quarelle|r(?:a(?:mco|b)|chi|my|pa|te|t)|s(?:da|ia|sociates)|t(?:hleta|torney)|u(?:ction|di(?:ble|o)?|spost|t(?:hor|os|o))|vianca|ws|xa|zure|[cdefgilmoqrstuwxz])|b(?:a(?:by|idu|n(?:a(?:mex|narepublic)|[dk])|r(?:c(?:elona|lay(?:card|s))|efoot|gains)|s(?:ket|e)ball|uhaus|yern|r)|b(?:va|[ct])|c[gn]|e(?:a(?:ts|uty)|er|ntley|rlin|st(?:buy)?|t)|harti|i(?:ble|ke|ngo?|[doz])|l(?:ack(?:friday)?|o(?:ckbuster|omberg|g)|ue)|m[sw]|npparibas|o(?:ats|ehringer|fa|nd|ok(?:ing)?|s(?:ch|t(?:ik|on))|utique|[motx])|r(?:adesco|idgestone|o(?:adway|ker|ther)|ussels)|u(?:ild(?:ers)?|siness|zz|y)|zh|[abdefghijmnorstvwyz])|c(?:a(?:fe|l(?:vinklein|l)|m(?:era|p)|non|p(?:etown|ital(?:one)?)|r(?:avan|ds|eers?|[es])|s(?:ino|[aeh])|t(?:ering|holic)|[blmrt])|b(?:re|[ans])|e(?:nter|rn|o)|f[ad]|h(?:a(?:nn?el|rity|se|t)|eap|intai|r(?:istmas|ome)|urch)|i(?:priani|rcle|sco|t(?:adel|ic|yeats|[iy]))|l(?:aims|eaning|i(?:ck|ni(?:que|c))|o(?:thing|ud)|ub(?:med)?)|o(?:ach|des|ffee|l(?:leg|ogn)e|m(?:cast|m(?:bank|unity)|p(?:a(?:ny|re)|uter)|sec)|n(?:dos|s(?:truction|ulting)|t(?:act|ractors))|o(?:king(?:channel)?|[lp])|rsica|u(?:ntry|pons?|rses)|m)|pa|r(?:edit(?:card|union)?|icket|own|uises?|s)|uisinella|y(?:mr|o)u|[acdfghiklmnoruvwxyz])|d(?:a(?:bur|nce|t(?:ing|sun|[ae])|[dy])|clk|ds|e(?:al(?:er|s)?|gree|l(?:ivery|oitte|ta|l)|mocrat|nt(?:al|ist)|si(?:gn)?|v)|hl|i(?:amonds|et|gital|rect(?:ory)?|s(?:co(?:unt|ver)|h)|y)|np|o(?:c(?:tor|s)|mains|wnload|[gt])|rive|tv|u(?:bai|nlop|pont|rban)|v(?:ag|r)|[ejkmoz])|e(?:a(?:rth|t)|co|d(?:eka|ucation|u)|m(?:ail|erck)|n(?:ergy|gineer(?:ing)?|terprises)|pson|quipment|r(?:icsson|ni)|s(?:tate|q)|tisalat|u(?:rovision|s)|vents|x(?:change|p(?:ert|osed|ress)|traspace)|[cegrstu])|f(?:a(?:ge|i(?:rwinds|th|l)|mily|ns|rm(?:ers)?|s(?:hion|t)|n)|e(?:dex|edback|rr(?:ari|ero))|i(?:at|d(?:elity|o)|lm|na(?:nc(?:ial|e)|l)|r(?:eston|mdal)?e|sh(?:ing)?|tness|t)|l(?:i(?:ckr|ghts|r)|o(?:rist|wers)|y)|o(?:o(?:dnetwork|tball|d)|r(?:ex|sale|um|d)|undation|[ox])|r(?:e(?:senius|e)|o(?:gans|nt(?:doo|ie)r)|l)|tr|u(?:jitsu|nd|rniture|tbol|n)|yi|[ijkmor])|g(?:a(?:ll(?:ery|up|o)|mes?|rden|[lpy])|biz|dn|e(?:nt(?:ing)?|orge|a)|gee|i(?:fts?|v(?:es|ing))|l(?:ass|ob(?:al|o)|e)|m(?:ail|bh|[ox])|o(?:daddy|l(?:dpoint|[df])|o(?:dyear|gle|g)|[optv])|r(?:a(?:inger|phics|tis)|een|ipe|o(?:cery|up))|u(?:ardian|cci|ge|i(?:de|tars)|ru)|[abdefghilmnpqrstuwy])|h(?:a(?:ir|mburg|ngout|us)|bo|dfc(?:bank)?|e(?:alth(?:care)?|l(?:sinki|p)|r(?:mes|e))|gtv|i(?:phop|samitsu|tachi|v)|kt|o(?:ckey|l(?:dings|iday)|me(?:depot|goods|sense|s)|nda|rse|s(?:pital|ting|t)|t(?:ele?s|mail)|use|[tw])|sbc|ughes|y(?:att|undai)|[kmnrtu])|i(?:bm|c(?:bc|[eu])|eee|fm|kano|m(?:amat|db|mo(?:bilien)?)|n(?:dustries|f(?:initi|o)|s(?:titut|ur(?:anc)?)e|t(?:ernational|uit)|vestments|[cgkt])|piranga|rish|s(?:maili|tanbul|t)|t(?:au|v)|[delmnoqrst])|j(?:a(?:guar|va)|cb|e(?:ep|tzt|welry)|io|ll|mp|nj|o(?:b(?:urg|s)|[ty])|p(?:morgan|rs)|u(?:egos|niper)|[emop])|k(?:aufen|ddi|erry(?:hotel|logistic|propertie)s|fh|i(?:ds|nd(?:er|le)|tchen|wi|[am])|o(?:eln|matsu|sher)|p(?:mg|n)|re?d|uokgroup|yoto|[eghimnprwyz])|l(?:a(?:caixa|m(?:borghini|er)|n(?:c(?:aster|ia)|drover|xess|d)|salle|t(?:ino|robe)|wyer|[tw])|ds|e(?:ase|clerc|frak|g(?:al|o)|xus)|gbt|i(?:dl|fe(?:insurance|style)?|ghting|ke|lly|m(?:ited|o)|n(?:coln|de|k)|psy|v(?:ing|e))|l[cp]|o(?:ans?|c(?:ker|us)|ndon|tt[eo]|ve|l)|pl(?:financial)?|tda?|u(?:ndbeck|x(?:ury|e))|[abcikrstuvy])|m(?:a(?:cys|drid|i(?:son|f)|keup|n(?:agement|go)|r(?:ket(?:ing|s)?|riott|shalls)|serati|ttel|[np])|ba|ckinsey|e(?:dia|et|lbourne|m(?:orial|e)|nu|rckmsd|[dn])|i(?:ami|crosoft|n[it]|tsubishi|[lt])|l[bs]|ma|o(?:bi(?:le)?|da|n(?:ash|ey|ster)|r(?:mon|tgage)|scow|to(?:rcycles)?|vie|[eimv])|sd|t[nr]|u(?:s(?:eum|ic)|tual)|[acdeghklmnopqrstuvwxyz])|n(?:a(?:goya|me|tura|vy|b)|ba|e(?:t(?:bank|flix|work)|ustar|ws|x(?:tdirect|us|t)|[ctw])|fl|go|hk|i(?:co|k(?:on|e)|nja|ssa[ny])|o(?:kia|rt(?:hwesternmutual|on)|w(?:ruz|tv)|w)|r[aw]|tt|yc|[acefgilopruz])|o(?:b(?:server|i)|ffice|kinawa|l(?:ayan(?:group)?|dnavy|lo)|mega|n(?:line|[egl])|oo|pen|r(?:a(?:cl|ng)e|ganic|igins|g)|saka|t(?:suka|t)|vh|m)|p(?:a(?:ge|nasonic|r(?:is|t(?:ners|[sy])|s)|ssagens|y)|ccw|et|fizer|h(?:armacy|ilips|o(?:ne|to(?:graphy|s)?)|ysio|d)|i(?:c(?:t(?:et|ures)|s)|n[gk]|oneer|zza|[dn])|l(?:a(?:ce|ystation|y)|u(?:mbing|s))|nc|o(?:hl|ker|litie|rn|st)|r(?:a(?:merica|xi)|ess|ime|o(?:ductions|gressive|mo|pert(?:ies|y)|tection|[df])|udential|[ou])|ub|wc|[aefghklmnrstwy])|q(?:pon|ue(?:bec|st)|a)|r(?:a(?:cing|dio)|e(?:a(?:l(?:estate|t(?:or|y))|d)|cipes|d(?:stone|umbrella)|hab|i(?:sen?|t)|liance|nt(?:als)?|p(?:air|ort|ublican)|st(?:aurant)?|views?|xroth|[dn])|i(?:c(?:hardli|oh|h)|[lop])|o(?:c(?:her|ks)|deo|gers|om)|svp|u(?:gby|hr|n)|we|yukyu|[eosuw])|s(?:a(?:arland|fe(?:ty)?|kura|l(?:on|e)|ms(?:club|ung)|n(?:dvik(?:coromant)?|ofi)|rl|ve|xo|[ps])|b[is]|c(?:h(?:aeffler|midt|o(?:larships|ol)|ule|warz)|ience|ot|[ab])|e(?:a(?:rch|t)|cur(?:ity|e)|ek|lect|ner|rvices|ven|xy|[wx])|fr|h(?:a(?:ngrila|rp|w)|ell|i(?:ksh)?a|o(?:es|pping|uji|wtime|[pw]))|i(?:lk|n(?:gles|a)|te)|k(?:in|ype|[iy])|ling|m(?:art|ile)|ncf|o(?:c(?:cer|ial)|ft(?:bank|ware)|hu|l(?:ar|utions)|n[gy]|y)|p(?:ace|or?t|a)|rl|t(?:a(?:da|ples|te(?:bank|farm)|r)|cgroup|o(?:ckholm|r(?:ag)?e)|ream|ud(?:io|y)|yle|c)|u(?:cks|pp(?:l(?:ies|y)|ort)|r(?:gery|f)|zuki)|w(?:atch|iss)|y(?:dney|stems)|[abcdeghijklmnorstuvxyz])|t(?:a(?:ipei|lk|obao|rget|t(?:a(?:motors|r)|too)|xi|[bx])|ci|dk|e(?:am|ch(?:nology)?|masek|nnis|va|l)|h(?:eat(?:er|re)|d)|i(?:aa|ckets|enda|ffany|ps|r(?:es|ol))|j(?:max)?x|kmaxx|mall|o(?:day|kyo|ols|ray|shiba|tal|urs|wn|y(?:ota|s)|p)|r(?:a(?:d(?:ing|e)|ining|vel(?:channel|ers(?:insurance)?)?)|ust|v)|u(?:be|nes|shu|i)|vs|[cdfghjklmnortvwz])|u(?:b(?:ank|s)|n(?:i(?:com|versity)|o)|ol|ps|[agksyz])|v(?:a(?:cations|n(?:guard|a))|e(?:gas|ntures|r(?:isign|sicherung)|t)|i(?:ajes|deo|king|llas|rgin|s(?:ion|a)|v[ao]|[gnp])|laanderen|o(?:dka|l(?:kswagen|vo)|t(?:ing|[eo])|yage)|uelos|[aceginu])|w(?:a(?:l(?:es|mart|ter)|ng(?:gou)?|tch(?:es)?)|e(?:ather(?:channel)?|b(?:cam|er|site)|dding|i(?:bo|r)|d)|hoswho|i(?:en|ki|lliamhill|n(?:dows|ners|e)|n)|me|o(?:lterskluwer|odside|r(?:ks|ld|k)|w)|t[cf]|[fs])|x(?:box|erox|finity|i(?:hua)?n|xx|yz)|y(?:a(?:chts|hoo|maxun|ndex)|o(?:dobashi|ga|kohama|utube|u)|un|[et])|z(?:a(?:ppos|ra)|ero|ip|one|uerich|[amw]))(?:\/\S*)?(?=\s|$)
```

I hope my thoughts help to fight spam a little :)

Reply:
Thank you for the review and feedback on this PR. I would appreciate your review and feedback on the latest commit (7207bf4). Tests based on this latest regex proposal can be found at https://regex101.com/r/PMJy6M/1

As I continue to consider this, one significant drawback to maintaining a regex that is based on published TLDs is the maintenance effort required to ensure the regex is updated appropriately as new TLDs are introduced. This becomes extra complicated when leveraging optimized regular expressions given the need to re-optimize.

In 7207bf4, I have considered a different approach and instead look to target (as you suggest) just those URLs that would otherwise load in a browser, sacrificing matches against obfuscated URLs with spaces, etc. I feel this is a fair tradeoff between being "perfect" and being "reasonable". This regular express also improves Unicode support.

The tradeoff - dot separated character strings with 2 or more characters per section are now matched. For example, "bar.lua" regresses from your regex to that proposed in 7207bf4 ("lua" not being a valid TLD), but I feel this is a fair tradeoff given the ease of maintenance effort for this approach.

With regard to overmatching script references such as terra.py / myscript.java / etc., the PR introduces an "exempted" set, where libraries that match certain strings can be explicitly exempted from the external link detection (https://github.com/terra-money/station/pull/202/files#diff-23d20fc3b5511f5f818ebaa3271c88da48e9deb5fc1e4a942e4f5a13905b10f7R33).

Again, thank you - I appreciate the effort toward a solution that fosters transparent governance, while providing reasonable safeguards to spam/scam proposals.

Reply:
I think we're close... great suggestions. Thank you for the work on IPv4/IPv6 detection. I made a few adjustments to your regex to address repeated punctuation characters at the end of a URL, and added additional test cases.

See https://regex101.com/r/rk4hTj/1 for an update test harness.

Here's my thoughts moving forward - no matter how hard we try, there will always be edge cases. I'd propose that the current regex as of 316de17 is powerful in accomplishing the goal of broadly disincentivizing spam/scam proposals. There may be additional work around exceptions and file extension matching, but given that proposal authors are notified prior to submitting governance proposals if URL detection is matched (and should really be offloading conversation regarding governance to agora or similar on-chain forums), I am comfortable with the current state of matching as an attempt to inform the community of potential spam/scam URLs.

Will gladly consider your thoughts, but I'm comfortable that we're close enough to bring value to interchain station given the changes proposed in 316de17.

Reply:
Great, I think we are really close. One last suggestion from my side would be to use `(?:\S*?:\/\/)?` instead of `(?:https?:\/\/)?` to match other protocols like HTTP or URLs like *://example.com* as well.

Apart from that everything looks good :)

Reply:
My concern with `(?:\S*?:\/\/)?` is that URLs referenced within parentheses (fairly common in gov props) will match the leading parentheses. For valid links to destinations such as agora, this would lead to broken links. I've updated to reduce the \S to the character set of official URI schemes https://en.wikipedia.org/wiki/List_of_URI_schemes with `(?:[a-z0-9-.]*?:\/\/)?`

Updated test bench at https://regex101.com/r/9J212C/1

Reply:
Good point, I didn't think of that. I would say that the regex cannot get any better at this point, so it is ready to merge.

Thanks for the fruitful discussion :)